cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh

```diff
@@ -41,176 +41,80 @@ namespace cuspatial {
 
-  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x;
-       idx < std::distance(linestring_points_first, thrust::prev(linestring_points_last));
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multilinestrings.num_points();
        idx += gridDim.x * blockDim.x) {
     // Search from the part offsets array to determine the part idx of current linestring point
-    auto linestring_part_offsets_iter = thrust::upper_bound(
-      thrust::seq, linestring_part_offsets_first, linestring_part_offsets_last, idx);
-
+    auto part_idx = multilinestrings.part_idx_from_point_idx(idx);
     // Pointer to the last point in the linestring, skip iteration.
     // Note that the last point for the last linestring is guarded by the grid-stride loop.
-    if (linestring_part_offsets_iter != linestring_part_offsets_last &&
-        *linestring_part_offsets_iter - 1 == idx) {
-      continue;
-    }
-
-    auto part_offsets_idx =
-      thrust::distance(linestring_part_offsets_first, thrust::prev(linestring_part_offsets_iter));
+    if (!multilinestrings.is_valid_segment_id(idx, part_idx)) continue;
 
-    // Search from the linestring geometry offsets array to determine the geometry idx of current
-    // linestring point
-    auto geometry_offsets_iter = thrust::upper_bound(thrust::seq,
-                                                     linestring_geometry_offset_first,
-                                                     linestring_geometry_offset_last,
-                                                     part_offsets_idx);
-    // geometry_idx is also the index to corresponding multipoint in the pair
-    auto geometry_idx =
-      thrust::distance(linestring_geometry_offset_first, thrust::prev(geometry_offsets_iter));
+    // Search from the linestring geometry offsets array to determine the geometry idx of
+    // current linestring point
+    auto geometry_idx = multilinestrings.geometry_idx_from_part_idx(part_idx);
 
     // Reduce the minimum distance between different parts of the multi-point.
-    vec_2d<T> const a      = linestring_points_first[idx];
-    vec_2d<T> const b      = linestring_points_first[idx + 1];
+    auto [a, b]            = multilinestrings.segment(idx);
     T min_distance_squared = std::numeric_limits<T>::max();
 
-    for (auto point_idx = point_geometry_offset_first[geometry_idx];
-         point_idx < point_geometry_offset_first[geometry_idx + 1];
-         point_idx++) {
-      vec_2d<T> const c = points_first[point_idx];
-
+    for (vec_2d<T> const& c : multipoints[geometry_idx]) {
       // TODO: reduce redundant computation only related to `a`, `b` in this helper.
       auto const distance_squared = point_to_segment_distance_squared(c, a, b);
-      min_distance_squared        = std::min(distance_squared, min_distance_squared);
+      min_distance_squared        = min(distance_squared, min_distance_squared);
     }
-
     atomicMin(&thrust::raw_reference_cast(*(distances + geometry_idx)),
```

Comment:
Can this be simplified? 

```suggestion
    atomicMin(&distances[geometry_idx],
```

Reply:
Yes*
But not without this support: https://github.com/isVoid/cuspatial/blob/17b99149ccfbf85bd2abfeeaf8329f2de70cfe02/cpp/include/cuspatial/detail/utility/device_atomics.cuh#L184-L186

Reply:
This is because `distances[geometry_idx]` can be many things. If it's a `float` then it's all good. Other times it can be `thrust::device_reference`, which would require some extra work to cast that into a raw pointer. The easiest of all approach is just to override atomic to accept `thrust::device_ptr`. Let's hope `atomic_ref` plays nicely with device vectors in the future.